File: day02/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	valid := 0
	valid2 := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())
		r := strings.Split(fields[0], "-")
		a, err := strconv.Atoi(r[0])
		if err != nil {
			log.Fatal(err)
		}
		b, err := strconv.Atoi(r[1])
		if err != nil {
			log.Fatal(err)
		}
		c := strings.TrimRight(fields[1], ":")[0]
		count := 0
		for _, i := range fields[2] {
			if i == rune(c) {
				count++
			}
		}
		if count >= a && count <= b {
			valid++
		}
		if fields[2][a-1] == c && fields[2][b-1] != c {
			valid2++
		} else if fields[2][a-1] != c && fields[2][b-1] == c {
			valid2++
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
	fmt.Println("Part 1:", valid)
	fmt.Println("Part 2:", valid2)
}

File: day02/src/main.rs
use regex::Regex;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let re = Regex::new(r"^(\d+)-(\d+) (a-z): ([a-z]+)$")?;

    let mut valid = 0;
    let mut valid2 = 0;

    for line in reader.lines() {
        let caps = re.captures(&line?)?;
        let a = caps.get(1)?.parse::<i32>()?;
        let b = caps.get(2)?.parse::<i32>()?;
        let c = caps.get(3)?.parse::<char>()?;
        let pass = caps.get(4)?;
        let count = pass.chars().map(|i| if i == c { 1 } else { 0 }).sum();
        if count >= a && count <= b {
            valid += 1;
        }
    }

    println!("Part 1: {}", valid);
    Ok(())
}
